apps/starknet_app/starknet_sign_txn.c

```diff
@@ -0,0 +1,627 @@
+static bool check_which_request(const starknet_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->sign_txn.which_request) {
+    starknet_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                        ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static void send_response(const pb_size_t which_response) {
+  starknet_result_t result = init_starknet_result(STARKNET_RESULT_SIGN_TXN_TAG);
+  result.sign_txn.which_response = which_response;
+  starknet_send_result(&result);
+}
+
+static bool validate_request_data(const starknet_sign_txn_request_t *request) {
+  bool status = true;
+
+  if (!starknet_derivation_path_guard(
+          request->initiate.derivation_path,
+          request->initiate.derivation_path_count)) {
+    starknet_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                        ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+  return status;
+}
+
+static bool handle_initiate_query(const starknet_query_t *query) {
+  char wallet_name[NAME_SIZE] = "";
+  char msg[100] = "";
+
+  if (!check_which_request(query, STARKNET_SIGN_TXN_REQUEST_INITIATE_TAG) ||
+      !validate_request_data(&query->sign_txn) ||
+      !get_wallet_name_by_id(query->sign_txn.initiate.wallet_id,
+                             (uint8_t *)wallet_name,
+                             starknet_send_error)) {
+    return false;
+  }
+
+  snprintf(msg,
+           sizeof(msg),
+           UI_TEXT_SIGN_TXN_PROMPT,
+           starknet_app.name,
+           wallet_name);
+  // Take user consent to sign transaction for the wallet
+  if (!core_confirmation(msg, starknet_send_error)) {
+    return false;
+  }
+
+  set_app_flow_status(STARKNET_SIGN_TXN_STATUS_CONFIRM);
+  memcpy(&starknet_txn_context->init_info,
+         &query->sign_txn.initiate,
+         sizeof(starknet_sign_txn_initiate_request_t));
+
+  send_response(STARKNET_SIGN_TXN_RESPONSE_CONFIRMATION_TAG);
+  // show processing screen for a minimum duration (additional time will add due
+  // to actual processing)
+  delay_scr_init(ui_text_processing, DELAY_SHORT);
+  return true;
+}
+
+static bool fetch_valid_input(starknet_query_t *query) {
+  if (!starknet_get_query(query, STARKNET_QUERY_SIGN_TXN_TAG) &&
+      !check_which_request(query, STARKNET_SIGN_TXN_REQUEST_TXN_TAG)) {
+    return false;
+  }
+
+  // Get txn type
+  starknet_txn_context->which_type = query->sign_txn.txn.which_type;
+  switch (starknet_txn_context->which_type) {
+    case STARKNET_SIGN_TXN_UNSIGNED_TXN_INVOKE_TXN_TAG: {
+      starknet_txn_context->invoke_txn = &query->sign_txn.txn.invoke_txn;
+    } break;
+
+    case STARKNET_SIGN_TXN_UNSIGNED_TXN_DEPLOY_TXN_TAG: {
+      starknet_txn_context->deploy_txn = &query->sign_txn.txn.deploy_txn;
+    } break;
+
+    default: {
+      // should not reach here;
+      return false;
+    }
+  }
+
+  if (1) {
```

Comment:
??

Reply:
😂 no idea why its there; remenants from previous implementation of app; fixed it